backend/proxy/proxy.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2019 Shift Devices AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package socksproxy
+
+import (
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+
+	"golang.org/x/net/proxy"
+)
+
+// SocksProxy holds the proxy address and wether to use it
+type SocksProxy struct {
+	useProxy     bool
+	proxyAddress string
+}
+
+// NewSocksProxy returns a new socks proxy instance
+func NewSocksProxy(useProxy bool, proxyAddress string) SocksProxy {
+	proxy := SocksProxy{
+		useProxy:     useProxy,
+		proxyAddress: proxyAddress,
+	}
+	return proxy
+}
+
+// GetTCPConnection returns a tcp connection. The connection is proxied, if useProxy is true.
+func (socksProxy *SocksProxy) GetTCPConnection(address string) (net.Conn, error) {
```

Comment:
A bit higher level:

`GetDialer(address string) (net.Dialer, error)`

then you don't need to do the dialing here.

Reply:
Ha, that's what I thought initially and was then surprised by the proxy dialer returning a dialer of its own type and not a net.Dialer. Only when connecting a generic net.Conn is returned. Using the proxy dialer directly would complicate the calling function.

Reply:
I thought it was the same too, but it could work by returning `proxy.Dialer`, which is an interface which `net.Dialer` implements

Reply:
That works, but then I cannot return an unproxied dialer.

Reply:
Why not? You can return `net.Dialer{}` which implements `proxy.Dialer` and is unproxied.

Reply:
net.Dialer does not implement proxy.Dialer. net.DIaler's Dial has a pointer receiver, while ([i.e. an instance of net.Dialer](https://golang.org/pkg/net/#Dialer.Dial)), while proxy.Dialer has [a non-pointer receiver](https://godoc.org/golang.org/x/net/proxy#Dial).

Reply:
Alright, `&net.Dialer{}` implements it then :smile: 

```Go
package main

import (
	"fmt"
	"net"

	"golang.org/x/net/proxy"
)

func main() {
	var dialer proxy.Dialer = &net.Dialer{}
	_, err := dialer.Dial("tcp", "golang.org:80")
	fmt.Println(err)
}
```

Reply:
I should stop reading documentation while im in some Ravine :P